File: number of islands.py
import sys
sys.setrecursionlimit(10**8)
from collections import deque


def isValid(rows, cols, row, col):
    return 0 <= row < rows and 0 <= col < cols


class Solution:

    def bfs(self, grid, visited, row, col):
        q = deque()
        # mask current cell as visited
        visited[row][col] = 1
        # since current cell is visited, append it in the queue
        q.append([row, col])
        directions = [[-1, 0], [-1, 1], [0, 1], [1, 1], [1, 0], [1, -1], [0, -1], [-1, -1]]
        while q:
            current = q.popleft()
            # explore all the 8 directions
            for direction in directions:
                next_row = current[0] + direction[0]
                next_col = current[1] + direction[1]
                if isValid(len(grid), len(grid[0]), next_row, next_col) and visited[next_row][next_col] != 1 and grid[next_row][next_col] == 1:
                    visited[next_row][next_col] = 1
                    q.append([next_row, next_col])

    def numIslands(self, grid):
        rows = len(grid)
        cols = len(grid[0])
        visited = [[0 for i in range(cols)] for j in range(rows)]
        numberOfIslands = 0
        for row in range(rows):
            for col in range(cols):
                if visited[row][col] == 0 and grid[row][col] == 1:
                    numberOfIslands += 1
                    self.bfs(grid, visited, row, col)

        return numberOfIslands

File: number of islands.cpp
bool isValidCell(int rows, int cols, int row, int col) {
        if (row >= 0 && row < rows && col >= 0 && col < cols) {
            return true;
        }
        return false;
    }
class Solution {
private:
    void bfs(vector<vector<char>>& grid, vector<vector<int>>& visited, int i, int j) {
        int rows = grid.size(), cols = grid[0].size();
        queue<pair<int, int>> q;
        // mark current position as visited
        visited[i][j] = 1;
        q.push({ i, j });
        vector<vector<int >> directions = { {-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}, {0, -1}, {-1, -1} };
        while (!q.empty()) {
            pair<int, int> current = q.front();
            q.pop();
            // explore all 8 directions
            for (auto & direction : directions) {
                int next_row = current.first + direction[0];
                int next_col = current.second + direction[1];
                // check if the next destination is already visited or not
                if (isValidCell(rows, cols, next_row, next_col) && !visited[next_row][next_col] && grid[next_row][next_col] == '1') {
                    visited[next_row][next_col] = 1;
                    q.push({ next_row, next_col });
                }
            }
        }
    }
public:
    // Function to find the number of islands.
    int numIslands(vector<vector<char>>& grid) {
        // Code here
        int rows = grid.size();
        int cols = grid[0].size();
        // visited array to keap track of visited cell
        vector<vector<int>> visited(rows, vector<int>(cols, 0));
        int numOfIslands = 0;
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                if (grid[row][col] == '1' && !visited[row][col]) {
                    bfs(grid, visited, row, col);
                    numOfIslands++;
                }
            }
        }
        return numOfIslands;
    }
};

//Space: O(n^2) and TC: O(n^2)[for exploring the whole grid] + O(n^2)[if the all the grid is marked as 1, then we have to search N^2 nodes for the first bfs]
